src/typechecker.erl

```diff
@@ -3423,6 +3431,47 @@ refinable(_) ->
 no_guards({clause, _, _, Guards, _}) ->
     Guards == [].
 
+%% Refines the types of bound variables using the assumption that a clause has
+%% mismatched.
+refine_vars_by_mismatching_clause({clause, _, Pats, Guards, _Block}, VEnv, TEnv) ->
+    PatternCantFail = are_patterns_matching_all_input(Pats, VEnv),
+    case Guards of
+        [[{call, _, {atom, _, Fun}, Args = [{var, _, Var}]}]] when PatternCantFail ->
+            %% Simple case: A single guard on the form `when is_TYPE(Var)'.
+            %% If Var was bound before the clause, which failed because of a
+            %% failing is_TYPE guard, we can exclude TYPE (as in is_TYPE).
+            case {VEnv, check_guard_call(Fun, Args)} of
+                {#{Var := Ty}, #{Var := GuardTy}} ->
+                    RefinedTy = try refine(Ty, GuardTy, TEnv)
+                                catch no_refinement -> Ty
+                                end,
+                    VEnv#{Var := RefinedTy};
+                _ ->
+                    %% No refinement
+                    VEnv
+            end;
+        _OtherGuards ->
+            %% No refinement
+            VEnv
```

Comment:
Great work. Would it be hard to add `andalso`/`and` and `orelse`/`or`?

Reply:
Possibly for in limited cases, but I'd prefer to leave it to a separate PR.

For a guard sequence like `is_atom(X); is_integer(X)`, where all the guards are on the same variable X, the type of X can be refined by excluding both atom() and integer().

Do you have any other example in mind?

Reply:
Totally fine to split it in multiple PRs.

This is a hair-pulling example a bit. Only useful if we want to have it as a test I guess.

```erlang
-spec test(integer() | atom() | list()) -> list().
test(X) ->
    if
        is_atom(X) orelse is_integer(X) -> [];
        true -> X
    end.
```

It also reminded me of the record part I did. Could also be another PR.

```erlang
-record(r1, {f :: integer()}).
-record(r2, {f :: integer()}).

-spec record(#r1{} | #r2{}) -> integer().
record(#r1{f = F}) -> F;
record(R2) -> R2#r2.f.
```